File: LeetCode/C++/5. Longest Palindromic Substring.cpp
/// LeetCode 5. Longest Palindromic Substring by C++
/// write by LazyWolfLin in 17.03.02

/// Manacher Algorithms
class Solution
{
public:
    string longestPalindrome(string s)
    {
        //initialize
        string t="$#";
        for (int i=0;i<s.size();++i)
        {
            t+=s[i];
            t+="#";
        }
        vector<int> f(t.size(),0);
        int current_center=0,current_right=0,ans_len=0,ans_center=0;
        //calculation
        for (int i=1;i<t.size();++i)
        {
            f[i]=current_right>i?min(f[2*current_center-i],current_right-i):1;
            while (t[i+f[i]]==t[i-f[i]]) ++f[i];
            if (i+f[i]>current_right)
            {
                current_right=i+f[i];
                current_center=i;
            }
            if (f[i]>ans_len)
            {
                ans_len=f[i];
                ans_center=i;
            }
        }
        return s.substr((ans_center-ans_len)/2,ans_len-1);
    }
};

File: LeetCode/Python3/5. Longest Palindromic Substring.py
# LeetCode 5. Longest Palindromic Substring by Python3
# write by LazyWolfLin in 18.08.02

# Manacher Algorithms
class Solution:
    def longestPalindrome(self, s):
        """
        :type s: str
        :rtype: str
        """
        t = "$#"
        for i in range(0, len(s)):
            t += s[i]
            t += "#"
        t += '!'
        f = [0]
        current_center = current_right = ans_len = ans_center = 0
        for i in range(1, len(t) - 1):
            f.append(min(f[2 * current_center - i], current_right - i) if current_right > i else 1)
            while t[i + f[i]] == t[i - f[i]]:
                f[i] += 1
            if i + f[i] > current_right:
                current_right = i + f[i]
                current_center = i
            if f[i] > ans_len:
                ans_len = f[i]
                ans_center = i
        return s[(ans_center - ans_len) // 2 : (ans_center + ans_len) // 2 - 1]